_posts/2017-03-17-getting-rest-right.md

```diff
@@ -0,0 +1,120 @@
+
+There are a couple of different ways you can go about this.
+
+1. In URL versioning (http://api.example.com/v1/article/43)
+   - Probably the most popular, but unfortunately semi-breaks the "URLs represent a resource" principle
+2. Custom header (X-API-Version: v1)
+   - Developers have to learn a new header just to use your API.
+3. Accept header (Accept: application/vnd.article-v1+json)
+   - Developers can't explore your API just using a browser.
+
+There's no easy answer, but I usually go for the compromise of in URL versioning for easy exploration of the API in combination with...
+
+## HATEOAS
+
+Standing for the easy to remember "Hypermedia As The Engine Of Application State", it's a fancy way of saying - return URLs to further resources in your API.
+
+If you have a request to /v1/article, it should return something similar to the following:
+
+```JSON
+GET /v1/article HTTP/1.1
+Host: api.example.com
+Accept: application/json
+
+[
+  {
+    "id": 42,
+    "name": "HG2G - Don't Panic",
+    "url": "/v1/article/42"
+  },
+  {
+    "id": 256,
+    "name": "Two to the Eight",
+    "url": "/v1/article/256"
+  }
+]
+```
+
+The url property of the article object allows the consumer to know how view the detail of the article object, without having to look at your documentation (you did ~~write~~ generate documentation right?). If you're using Postman, you can just click on the link and it will take you straight to the next resource!
+
+## SSL Everywhere
+
+With the invention of Let's Encrypt, there is no longer any excuse not to secure your API endpoint.
+
+If you've got HTTPS enabled, and you receive a request to HTTP - just throw an error such as `400 Bad Request {"error": "HTTPS Required"}` rather than redirecting to HTTPS.
+
+## Errors and HTTP status codes
+
+There are more [HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) than you can shake a stick at. Also all good libraries should be able to read these, so you may as well utilise them.
+
+* 2xx series - success
+* 3xx series - redirection
+* 4xx series - client errors
+* 5xx series - server errors
+
+Using these errors, especially for success can avoid you having to use an envelope in your response. If you're returning a list of articles and everything was fine, you can just 200 OK, and return the list. There's no need for:
+
+```JSON
+{
+  "status": 200,
+  "status_msg": "OK",
+  "articles": [
+    {
+      "id": 42,
+      "name": "HG2G - Don't Panic",
+      "url": "/v1/article/42"
+    }
+    ...
+  ]
+}
+```
+
+## Caching
+
+I'm talking about both types of caching, server side and client side. 
+
+The server should tell the client if they need to get new data. You can use `Last-Modified` or `Etags` headers for this.
+
+Also the server should keep a cache on the data from the datastore, especially if it's immutable or you when the data expires. This will keep response times snappy and consumers happy!
```

Comment:
`of the data` instead of `on the data`?
`especially if it's immutable or you when the data expires`. I think that there's something missing here.

Finally, the viability of caching is use-case dependent and using it can be more trouble than it's worth, so I'd probably avoid a strong statement that data _should_ be cached.